content/things/webgpu.md

```diff
@@ -0,0 +1,586 @@
+
+With the advent of ML, neural networks, and dare I say cryptocurrencies, GPUs have shown that they can be useful for more than just drawing triangles on the screen. Using GPUs for calculations of any kind is often called General-Purpose GPU or GPGPU, and WebGL 1 is not great at that. If you wanted to process arbitrary data to the GPU you have to encode it as a texture, decode it in a shader, do your calculations and then re-encode the result as a texture. WebGL 2 made this a lot easier with [Transform Feedback], but WebGL2 wasn’t supported in Safari until September 2021, so it wasn’t really an option (while most other browers supported WebGL2 since January 2017).
+
+## WebGPU
+
+Outside of the web, a new generation of graphics APIs have established themselves which expose a more low-level interface to graphics cards and accommodate multiple requirements. On the one hand, GPUs are almost ubiquitous now. Even our mobile devices have capable GPUs built in. As a result, _both_ modern graphics programming (like ray tracing) and GPGPU use-cases are increasingly common. On the other hand, most of our devices have multi-core processors, so being able to interact with the GPU from multiple threads can be an important optimization vector. While they were at it, they also revisited some design decisions of the previous APIs and front-loaded a lot of the validation work that GPUs have to do, allowing the developer to squeeze more performance out of the GPUs.
+
+The most popular of the next-gen GPU APIs are Vulcan by the Khronos Group, Metal by Apple and DirectX 12 by Microsoft. To bring these new capabilities to the web, WebGPU was born. It abstracts the idiosyncrasies of these different APIs and has been standardized in the W3C with all major browser vendors having a seat at the table. Due to their low-level nature and their sheer power, WebGPU has a bit of a learning curve and is relatively heavy on the setup, but I’ll try to break it down as best I can. As the title of this blog post implies, I will restrict myself to GPGPU use-case without using WebGPU to draw pixels. I think it may make it easier to grasp as an introduction to WebGPU.
```

Comment:
If you don't mind, it feels like it would be worthwhile here to say something high-level about why WebGPU doesn't directly mimic any of the named APIs in the same way that WebGL did. Something along the lines of:

> Unlike WebGL, which mimicked it's native API counterpart very closely, WebGPU isn't a direct wrapper over any of these native APIs. This is partially because no single API has native drivers on every system the web platform targets, but also because many of the patterns present in the native APIs, such as extremely low-level memory management, aren't appropriate for a web-facing API. As a result, the choice was made to create a new API that could expose the capabilities of these modern APIs in a web-friendly form: WebGPU. It abstracts the idosyncrasies..."